app_dart/test/service/luci_build_service_test.dart

```diff
@@ -485,140 +487,138 @@ void main() {
     });
 
-    test('Do not rerun a flutter Mac builder with infra failure: -9 in shards but not supported builder', () async {
+    test('Do not rerun a successful flutter builder', () async {
+      config.db.values[totCommit.key] = totCommit;
       config.maxLuciTaskRetriesValue = 1;
       const LuciTask luciTask = LuciTask(
           commitSha: 'abc',
           ref: 'refs/heads/master',
-          status: Task.statusInfraFailure,
+          status: Task.statusSucceeded,
           buildNumber: 1,
-          builderName: 'Mac abc',
-          summaryMarkdown:
-              'recipe infra failure: Infra Failure: Step(\'display builds.build(s) failed\') (retcode: 1)');
+          builderName: 'Mac abc');
       final bool rerunFlag = await service.checkRerunBuilder(
-        commitSha: 'abc',
+        commit: commit,
         luciTask: luciTask,
         retries: 0,
         repo: 'flutter',
+        datastore: datastore,
       );
       expect(rerunFlag, false);
     });
 
-    test('Do not rerun a flutter Mac builder with non -9 retcode', () async {
+    test('Do not rerun a non-supported repo builder', () async {
+      commit.repository = 'flutter/engine';
+      config.db.values[totCommit.key] = totCommit;
       config.maxLuciTaskRetriesValue = 1;
       const LuciTask luciTask = LuciTask(
           commitSha: 'abc',
           ref: 'refs/heads/master',
-          status: Task.statusFailed,
+          status: Task.statusInfraFailure,
           buildNumber: 1,
-          builderName: 'Mac abc',
-          summaryMarkdown: 'test failure');
+          builderName: 'Mac abc');
       final bool rerunFlag = await service.checkRerunBuilder(
-        commitSha: 'abc',
+        commit: commit,
         luciTask: luciTask,
         retries: 0,
-        repo: 'flutter',
+        repo: 'engine',
+        datastore: datastore,
       );
       expect(rerunFlag, false);
     });
 
-    test('Do not rerun a flutter non-Mac builder', () async {
+    test('Do not rerun a flutter builder exceeding retry limit', () async {
+      config.db.values[totCommit.key] = totCommit;
       config.maxLuciTaskRetriesValue = 1;
       const LuciTask luciTask = LuciTask(
           commitSha: 'abc',
           ref: 'refs/heads/master',
           status: Task.statusInfraFailure,
           buildNumber: 1,
-          builderName: 'Linux abc',
-          summaryMarkdown: 'Step(\'checkout source code.Checkout flutter/flutter.git fetch\') (retcode: -9)');
+          builderName: 'Mac abc');
       final bool rerunFlag = await service.checkRerunBuilder(
-        commitSha: 'abc',
+        commit: commit,
         luciTask: luciTask,
-        retries: 0,
+        retries: 1,
         repo: 'flutter',
+        datastore: datastore,
       );
       expect(rerunFlag, false);
     });
 
-    test('Do not rerun a flutter Mac builder exceeding attempt limit', () async {
+    test('Do not rerun a flutter builder when not blocking the tree', () async {
```

Comment:
Do we need at least 3 commits and more than two tasks to make this test relevant? e.g. oldest fails, the one in the middle fails with a different task, etc?

Reply:
Not necessary. We cares only the TOT commit: if the current failed task comes from the TOT commit, then it is blocking the tree and we need to rerun; if the task is not from the TOT commit, then we would expect there is a newer commit on top of this one and the same task of the newer commit would potentially help open the tree, and we will not rerun this task.
To make it more clear, updated to add both the current commit and a TOT commit to the fake datastore.